File: Python/src/mqtt/mqtt_tree_structure.py
from typing import List, Dict, Optional, Iterator, Any, Union
from collections.abc import Iterable

from mqtt.mqtt_handlers.mqtt_handler import MqttHandler

class MqttTreeStructure(Iterable):
    def __init__(self, topic: str):
        self.topic = topic
        self.handlers: List[MqttHandler] = []
        self.parent: Optional['MqttTreeStructure'] = None
        self.children: Dict[str, 'MqttTreeStructure'] = {}

    def get_child(self, topic: str) -> 'MqttTreeStructure':
        return self.children[topic]

    def get_children(self) -> List['MqttTreeStructure']:
        return list(self.children.values())

    def __iter__(self) -> Iterator['MqttTreeStructure']:
        return iter(self.children.values())

    def count(self) -> int:
        return len(self.children)

    def add_handler(self, handler: MqttHandler):
        self.handlers.append(handler)

    def remove_handler(self, handler: MqttHandler):
        self.handlers.remove(handler)

    def get_handlers(self) -> List[MqttHandler]:
        return self.handlers

    def get_root(self) -> 'MqttTreeStructure':
        root = self
        while root.parent is not None:
            root = root.parent
        return root

    def get_tree_node(self, topic: str) -> Optional['MqttTreeStructure']:
        root = self.get_root()
        topics = topic.split('/')

        if not topics:
            return None

        current = root

        for t in topics:
            if t in current.children:
                current = current.get_child(t)
            else:
                return None
            
        return current

    def add_tree_node(self, topic: str, handler: Optional[MqttHandler] = None):
        topics = topic.split('/')
        
        if not topics:
            return

        current = self

        for t in topics:
            if t in current.children:
                current = current.get_child(t)
            else:
                new_node = MqttTreeStructure(t)
                new_node.parent = current
                current.children[t] = new_node
                current = new_node

        if handler:
            current.add_handler(handler)

    def get_mqtt_handlers(self, topic: str) -> List[MqttHandler]:
        root = self.get_root()
        topics = topic.split('/')

        if not topics:
            return None

        handlers = []
        current = [root]

        for t in topics:
            next_nodes = []

            for c in current:
                if t in c.children:
                    next_nodes.append(c.get_child(t))
                if "+" in c.children:
                    next_nodes.append(c.get_child("+"))
                if "#" in c.children:
                    handlers.extend(c.get_child("#").get_handlers())

            current = next_nodes

        for c in current:
            handlers.extend(c.get_handlers())

        return handlers

    def get_all_subscriptions(self, node: Optional['MqttTreeStructure']) -> List[str]:
        subscriptions = []
        tree_node = node or self.get_root()

        for key, child in tree_node.children.items():
            if child.handlers:
                subscriptions.append(key)
            
            if child.count() > 0:
                child_subscriptions = child.get_all_subscriptions(child)

                for child_sub in child_subscriptions:
                    subscriptions.append(f"{key}/{child_sub}")

        return subscriptions

File: Dotnet/Dotnet.Mqtt/MqttTreeStructure.cs
using System.Collections;

namespace Dotnet.Mqtt;

public class MqttTreeStructure(string topic) : IEnumerable<MqttTreeStructure>
{
    private readonly Dictionary<string, MqttTreeStructure> children = [];

    public readonly string Topic = topic;
    public List<IMqttHandler> Handlers = [];
    public MqttTreeStructure? Parent { get; set; }

    public MqttTreeStructure GetChild(string topic)
    {
        return children[topic];
    }

    public List<MqttTreeStructure> GetChildren()
    {
        return [.. children.Values];
    }

    public IEnumerator<MqttTreeStructure> GetEnumerator()
    {
        return children.Values.GetEnumerator();
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
        return GetEnumerator();
    }

    public int Count
    {
        get { return children.Count; }
    }

    public void AddHandler(IMqttHandler handler)
    {
        Handlers.Add(handler);
    }

    public void RemoveHandler(IMqttHandler handler)
    {
        Handlers.Remove(handler);
    }

    public List<IMqttHandler> GetHandlers()
    {
        return Handlers;
    }

    public MqttTreeStructure GetRoot()
    {
        var root = this;
        var isRoot = false;

        while (!isRoot)
        {
            if (root.Parent != null)
            {
                root = root.Parent;
            }
            else
            {
                isRoot = true;
            }
        }

        return root;
    }

    //Return the tree node according to the topic. There is always 0 or 1 path.
    public MqttTreeStructure? GetTreeNode(string topic)
    {
        var root = GetRoot();

        //split topic
        var topics = topic.Split('/');

        if (topics.Length == 0)
        {
            return null;
        }

        //Move through the tree according to the topic and return the treeNode.
        var current = root;

        foreach (var t in topics)
        {
            if (current.children.ContainsKey(t))
            {
                current = current.GetChild(t);
            }
            else
            {
                return null;
            }
        }

        return current;
    }

    public void AddTreeNode(string topic, IMqttHandler? handler)
    {
        var topics = topic.Split('/');

        if (topics.Length == 0)
        {
            return;
        }

        var current = this;

        foreach (var t in topics)
        {
            if (current.children.ContainsKey(t))
            {
                current = current.GetChild(t);
            }
            else
            {
                var newNode = new MqttTreeStructure(t)
                {
                    Parent = current
                };
                
                current.children.Add(t, newNode);
                current = newNode;
            }
        }

        if (handler != null)
        {
            current.AddHandler(handler);
        }
    }

    //Return all handlers that are subscribed to the topic
    public List<IMqttHandler>? GetMqttHandlers(string topic)
    {
        MqttTreeStructure root = GetRoot();

        var topics = topic.Split('/');

        if (topics.Length == 0)
        {
            return null;
        }

        List<IMqttHandler> handlers = [];
        List<MqttTreeStructure> current = [root];

        foreach (var t in topics)
        {
            List<MqttTreeStructure> next = [];

            foreach (var c in current)
            {
                if (c.children.ContainsKey(t))
                {
                    next.Add(c.GetChild(t));
                }

                //skip 1 level
                if (c.children.ContainsKey("+"))
                {
                    next.Add(c.GetChild("+"));
                }

                //endpoint (you will never have ../#/... as topic)
                if (c.children.ContainsKey("#"))
                {
                    handlers.AddRange(c.GetChild("#").GetHandlers());
                }
            }

            current = next;
        }

        current.ForEach(c => handlers.AddRange(c.GetHandlers()));

        return handlers;
    }

    public IEnumerable<string> GetAllSubscriptions(MqttTreeStructure? node)
    {
        List<string> subscriptions = [];

        var treeNode = node ?? GetRoot();

        foreach (var child in treeNode.children)
        {
            //only add subscriptions that have handlers
            if(child.Value.Handlers.Count > 0)
            {
                subscriptions.Add(child.Key);
            }

            if (child.Value.Count > 0)
            {
                var childSubscriptions = child.Value.GetAllSubscriptions(child.Value);

                foreach (var childSubscription in childSubscriptions)
                {
                    subscriptions.Add(child.Key + "/" + childSubscription);
                }
            }
        }

        return subscriptions;
    }
}